src/qudi/interface/data_instream_interface.py

```diff
@@ -20,397 +20,251 @@
 If not, see <https://www.gnu.org/licenses/>.
 """
 
+__all__ = ['StreamingMode', 'SampleTiming', 'DataInStreamConstraints', 'DataInStreamInterface']
+
 import numpy as np
+from typing import Union, Type, Iterable, Mapping, Optional, Dict, List, Tuple, Sequence
 from enum import Enum
 from abc import abstractmethod
 from qudi.core.module import Base
 from qudi.util.constraints import ScalarConstraint
 
 
-class DataInStreamInterface(Base):
-    """
-    Interface for a generic input stream of data points with fixed sampling rate and data type.
+class StreamingMode(Enum):
+    INVALID = -1
+    CONTINUOUS = 0
+    FINITE = 1
 
-    You can choose if a preset number of samples is recorded and buffered for read or if samples
-    are acquired continuously into a (circular) read buffer.
-    """
-    @property
-    @abstractmethod
-    def sample_rate(self):
-        """
-        The currently set sample rate
 
-        @return float: current sample rate in Hz
-        """
-        pass
+class SampleTiming(Enum):
+    INVALID = -1
+    CONSTANT = 0
+    TIMESTAMP = 1
+    RANDOM = 2
 
-    @property
-    @abstractmethod
-    def data_type(self):
-        """
-        Read-only property.
-        The data type of the stream data. Must be numpy type.
 
-        @return type: stream data type (numpy type)
-        """
-        pass
+class DataInStreamConstraints:
+    """ Collection of constraints for hardware modules implementing DataInStreamInterface """
+    def __init__(self,
+                 channel_units: Mapping[str, str],
+                 sample_timing: Union[SampleTiming, int],
+                 streaming_modes: Iterable[Union[StreamingMode, int]],
+                 data_type: Union[Type[int], Type[float], Type[np.integer], Type[np.floating]],
+                 channel_buffer_size: Optional[ScalarConstraint],
+                 sample_rate: Optional[ScalarConstraint] = None):
+        if not isinstance(sample_rate, ScalarConstraint) and sample_rate is not None:
+            raise TypeError(
+                f'"sample_rate" must be None or'
+                f'{ScalarConstraint.__module__}.{ScalarConstraint.__qualname__} instance'
+            )
+        if not isinstance(channel_buffer_size, ScalarConstraint):
+            raise TypeError(
+                f'"channel_buffer_size" must be '
+                f'{ScalarConstraint.__module__}.{ScalarConstraint.__qualname__} instance'
+            )
+        self._channel_units = {**channel_units}
+        self._sample_timing = SampleTiming(sample_timing)
+        self._streaming_modes = [StreamingMode(mode) for mode in streaming_modes]
+        self._data_type = np.dtype(data_type).type
+        self._channel_buffer_size = channel_buffer_size
+        if sample_rate is None:
+            if SampleTiming.CONSTANT in self._sample_timings:
+                raise ValueError('"sample_rate" ScalarConstraint must be provided if '
+                                 'SampleTiming.CONSTANT is permitted')
```

Comment:
I think there is a bug here: the attribute `self._sample_timings` does not exist. There is only a single `self._sample_timing`. The text of the error should also be changed accordingly I believe.

Reply:
Absolutely right, thank you. Should be fixed now.
Only `SampleTiming.RANDOM` does not require a `sample_rate` constraint. For `SampleTiming.TIMESTAMP` the set `sample_rate` is not assumed to be exact but still used to set the general pace of data acquisition.